File: spark-in-action/scala/src/main/scala/deployment/PiComputeClusterApp.scala
package deployment

import java.util.ArrayList
import org.apache.spark.api.java.function.{MapFunction, ReduceFunction}
import org.apache.spark.sql.{Encoders, Row, SparkSession}

object PiComputeClusterApp {
  private var counter = 0

  @SerialVersionUID(38446L)
  final private class DartMapper extends MapFunction[Row, Integer] {
    @throws[Exception]
    override def call(r: Row): Integer = {
      val x: Double = Math.random * 2 - 1
      val y: Double = Math.random * 2 - 1
      counter += 1
      if (counter % 100000 == 0)
        print("" + counter + " darts thrown so far")
      if (x * x + y * y <= 1) 1
      else 0
    }
  }

  // Reducer class, reduces the map of dots
  @SerialVersionUID(12859L)
  final private class DartReducer extends ReduceFunction[Integer] {
    override def call(x: Integer, y: Integer): Integer = x + y
  }

  def main(args: Array[String]): Unit = {

    val slices = 10
    val numberOfThrows = 100000 * slices
    println("About to throw " + numberOfThrows + " darts")

    val t0 = System.currentTimeMillis
    val spark = SparkSession.builder
      .appName("Spark Pi on a cluster")
      .master("spark://spark-master:7077")
      .config("spark.executor.memory", "512m")
      // Spark can share a jar from which it is launched - either via spark-submit or
      // via direct connection
      .getOrCreate

    val t1 = System.currentTimeMillis
    println("Session initialized in " + (t1 - t0) + " ms")

    val numsList = new ArrayList[Integer](numberOfThrows)

    // For Spark Encoder implicits
    import spark.implicits._

    for (i <- 1 to numberOfThrows)
      numsList.add(i)

    val incrementalDf = spark.createDataset(numsList).toDF
    val t2 = System.currentTimeMillis
    println("Initial dataframe built in " + (t2 - t1) + " ms")

    val dartsDs = incrementalDf.map(new DartMapper, Encoders.INT)
    val t3 = System.currentTimeMillis
    println("Throwing darts done in " + (t3 - t2) + " ms")

    val dartsInCircle = dartsDs.reduce(new DartReducer)
    val t4 = System.currentTimeMillis
    println("Analyzing results in " + (t4 - t3) + " ms")

    println("Pi is roughly " + 4.0 * dartsInCircle / numberOfThrows)

    spark.stop()
  }
}

File: spark-in-action/java/src/main/java/deployment/PiComputeClusterApp.java
package deployment;

import org.apache.spark.api.java.function.MapFunction;
import org.apache.spark.api.java.function.ReduceFunction;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Encoders;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

import java.util.ArrayList;
import java.util.List;

public class PiComputeClusterApp {
    private static final long serialVersionUID = -1546L;
    private static long counter = 0;

    private final class DartMapper implements MapFunction<Row, Integer> {
        private static final long serialVersionUID = 38446L;

        @Override
        public Integer call(Row r) throws Exception {
            double x = Math.random() * 2 - 1;
            double y = Math.random() * 2 - 1;
            counter ++;
            if (counter % 1000 == 0) {
                System.out.println("" + counter + " operations done so far");
            }
            return (x * x + y * y <= 1) ? 1 : 0;
        }
    }

    private final class DartReducer implements ReduceFunction<Integer> {
        private static final long serialVersionUID = 12859L;

        @Override
        public Integer call(Integer x, Integer y) {
            return x + y;
        }
    }

    public static void main(String[] args) {
        PiComputeClusterApp app = new PiComputeClusterApp();
        app.start(10);
    }

    public void start(int slices) {
        int numberOfThrows = 100000 * slices;
        System.out.println("About to throw " + numberOfThrows + " darts");

        long t0 = System.currentTimeMillis();
        // The session resides on the cluster manager.
        SparkSession spark = SparkSession.builder()
                .appName("JavaSparkPi on a cluster")
                .master("spark://un:7077")
                .config("spark.executor.memory", "4g")
                .config("spark.jars", "jar/path")
                .getOrCreate();

        long t1 = System.currentTimeMillis();
        System.out.println("Session initialized in " + (t1 - t0) + " ms");

        List<Integer> l = new ArrayList<>(numberOfThrows);
        for (int i = 0; i < numberOfThrows; i++) {
            l.add(i);
        }

        // The first dataframe is created in the executor
        Dataset<Row> incrementalDf = spark.createDataset(l, Encoders.INT()).toDF();
        long t2 = System.currentTimeMillis();
        System.out.println("Initial dataframe built in " + (t2 - t1) + " ms");

        // This step is added to the DAG, which sits in the cluster manager
        Dataset<Integer> dartsDs = incrementalDf.map(new DartMapper(), Encoders.INT());
        long t3 = System.currentTimeMillis();
        System.out.println("Throwing darts done in " + (t3 - t2) + " ms");

        // The result of the reduce operation is brought back to the application
        int dartsInCircle = dartsDs.reduce(new DartReducer());
        long t4 = System.currentTimeMillis();
        System.out.println("Analyzing result in " + (t4 - t3) + " ms");

        System.out.println("Pi is roughly " + 4.0 * dartsInCircle / numberOfThrows);

        spark.stop();

    }
}